File: clients/CSharp/Model/ServerMessage.cs
namespace Aicup2020.Model
{
    public abstract class ServerMessage
    {
        public abstract void WriteTo(System.IO.BinaryWriter writer);
        public static ServerMessage ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case GetAction.TAG:
                    return GetAction.ReadFrom(reader);
                case Finish.TAG:
                    return Finish.ReadFrom(reader);
                case DebugUpdate.TAG:
                    return DebugUpdate.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        public class GetAction : ServerMessage
        {
            public const int TAG = 0;
            public Model.PlayerView PlayerView { get; set; }
            public bool DebugAvailable { get; set; }
            public GetAction() {}
            public GetAction(Model.PlayerView playerView, bool debugAvailable)
            {
                this.PlayerView = playerView;
                this.DebugAvailable = debugAvailable;
            }
            public static new GetAction ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new GetAction();
                result.PlayerView = Model.PlayerView.ReadFrom(reader);
                result.DebugAvailable = reader.ReadBoolean();
                return result;
            }
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                PlayerView.WriteTo(writer);
                writer.Write(DebugAvailable);
            }
        }

        public class Finish : ServerMessage
        {
            public const int TAG = 1;
            public Finish() {}
            public static new Finish ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Finish();
                return result;
            }
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
            }
        }

        public class DebugUpdate : ServerMessage
        {
            public const int TAG = 2;
            public Model.PlayerView PlayerView { get; set; }
            public DebugUpdate() {}
            public DebugUpdate(Model.PlayerView playerView)
            {
                this.PlayerView = playerView;
            }
            public static new DebugUpdate ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new DebugUpdate();
                result.PlayerView = Model.PlayerView.ReadFrom(reader);
                return result;
            }
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                PlayerView.WriteTo(writer);
            }
        }
    }
}

File: clients/FSharp/Model/ServerMessage.fs
#nowarn "0058"
namespace Aicup2020.Model

type ServerMessageGetAction = {
    PlayerView: PlayerView;
    DebugAvailable: bool;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        this.PlayerView.writeTo writer
        writer.Write this.DebugAvailable
    static member readFrom(reader: System.IO.BinaryReader) = {
        PlayerView = PlayerView.readFrom reader
        DebugAvailable = reader.ReadBoolean()
    }

type ServerMessageFinish = struct end with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
    static member readFrom(reader: System.IO.BinaryReader) = new ServerMessageFinish()

type ServerMessageDebugUpdate = {
    PlayerView: PlayerView;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 2
        this.PlayerView.writeTo writer
    static member readFrom(reader: System.IO.BinaryReader) = {
        PlayerView = PlayerView.readFrom reader
    }
type ServerMessage = 
    | GetAction of ServerMessageGetAction
    | Finish of ServerMessageFinish
    | DebugUpdate of ServerMessageDebugUpdate
    with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | GetAction value -> value.writeTo writer
            | Finish value -> value.writeTo writer
            | DebugUpdate value -> value.writeTo writer
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> GetAction (ServerMessageGetAction.readFrom reader)
            | 1 -> Finish (ServerMessageFinish.readFrom reader)
            | 2 -> DebugUpdate (ServerMessageDebugUpdate.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)
